File: client/js/ophion.js
var _ = require('underscore');


require('es6-promise').polyfill();
require('isomorphic-fetch');
var callsync = require('./callsync')

function value(x) {
  var v = null
  if (_.isString(x)) {
    v = {'s': x}
  } else if (_.isNumber(x)) {
    if (x === Math.floor(x)) {
      v = {'n': x}
    } else {
      v = {'r': x}
    }
  }

  return v ? v : x
}


function OphionQuery(parent) {
  var parent = parent

  function labels(l) {
    if (!l) {
      l = []
    } else if (_.isString(l)) {
      l = [l]
    } else if (!_.isArray(l)) {
      console.log("not something we know how to make labels out of:")
      console.log(l)
    }

    return {'labels': l}
  }

  function by(b) {
    if (_.isString(b)) {
      return {'key': b}
    } else {
      return {'query': b.query}
    }
  }

  return {
    query: [],

    V: function(l) {
      this.query.push({'V': labels(l)})
      return this
    },

    E: function(l) {
      this.query.push({'E': labels(l)})
      return this
    },

    searchVertex: function(term, search) {
      if (search) {
        this.query.push({'searchVertex': {'term': term, 'search': search}})
      } else {
        this.query.push({'searchVertex': {'search': term}})
      }

      return this
    },

    incoming: function(l) {
      this.query.push({'in': labels(l)})
      return this
    },

    outgoing: function(l) {
      this.query.push({'out': labels(l)})
      return this
    },

    inEdge: function(l) {
      this.query.push({'inEdge': labels(l)})
      return this
    },

    outEdge: function(l) {
      this.query.push({'outEdge': labels(l)})
      return this
    },

    inVertex: function(l) {
      this.query.push({'inVertex': labels(l)})
      return this
    },

    outVertex: function(l) {
      this.query.push({'outVertex': labels(l)})
      return this
    },

    identity: function() {
      this.query.push({'identity': true})
      return this
    },

    mark: function(l) {
      this.query.push({'as': labels(l)})
      return this
    },

    select: function(l) {
      this.query.push({'select': labels(l)})
      return this
    },

    by: function(key) {
      // a key is either a string or an inner query that has already been
      // built and passed in.
      this.query.push({'by': _.isString(key) ? {'key': key} : {'query': key.query}})
      return this
    },

    id: function() {
      this.query.push({'id': true})
      return this
    },

    label: function() {
      this.query.push({'label': true})
      return this
    },

    values: function(l) {
      this.query.push({'values': labels(l)})
      return this
    },

    properties: function(l) {
      this.query.push({'properties': labels(l)})
      return this
    },

    propertyMap: function(l) {
      this.query.push({'propertyMap': labels(l)})
      return this
    },

    dedup: function(l) {
      this.query.push({'dedup': labels(l)})
      return this
    },

    limit: function(l) {
      this.query.push({'limit': l})
      return this
    },

    order: function(o, asc) {
      if (asc === undefined) {
        asc = true;
      }
      this.query.push({'order': {'key': o, 'ascending': asc}})
      return this
    },

    range: function(lower, upper) {
      this.query.push({'range': {'lower': lower, 'upper': upper}})
      return this
    },

    count: function() {
      this.query.push({'count': true})
      return this
    },

    path: function() {
      this.query.push({'path': true})
      return this
    },

    aggregate: function(label) {
      this.query.push({'aggregate': label})
      return this
    },

    // group: function(by) {
    //   if (!)
    //   query.push({'groupCount': label})
    //   return this
    // },

    groupCount: function(b) {
      this.query.push({'groupCount': by(b)})
      return this
    },

    is: function(condition) {
      this.query.push({'is': condition})
      return this
    },

    has: function(key, h) {
      var out = {'key': key}
      if (_.isString(h) || _.isNumber(h)) {
        out['value'] = value(h)
      } else if (_.isArray(h)) {
        out['query'] = h.query
      } else {
        out['condition'] = h
      }

      this.query.push({'has': out})
      return this
    },

    hasLabel: function(l) {
      this.query.push({'hasLabel': labels(l)})
      return this
    },

    hasNot: function(key) {
      this.query.push({'hasNot': key})
      return this
    },

    match: function(queries) {
      this.query.push({'match': {'queries': _.map(queries, function(query) {return query.query})}})
      return this
    },

    execute: function(callback) {
      parent.execute(this.query, callback)
    },

    run: function() {
      return parent.run(this.query)
    }
  }
}

function parseJson(s) {
  try {
    return JSON.parse(s)
  } catch(err) {
    return s
  }
}

function Ophion(site='') {
  var queryBase = '/vertex/query'

  return {
    run: function(query) {
      return fetch(site + queryBase, {
        method: 'POST',
        headers: {'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: JSON.stringify(query),
      }).then(function(response) {
        return response.text()
      }).then(function(text) {
        var lines = text.replace(/^\s+|\s+$/g, '').split("\n")
        return lines.map(parseJson)
      })
    },

    execute: function(query, callback) {
      fetch(site + queryBase, {
        method: 'POST',
        headers: {'Content-Type': 'application/json', 'Accept': 'application/json'},
        body: JSON.stringify(query),
      }).then(function(response) {
        return response.text()
      }).then(function(text) {
        var lines = text.replace(/^\s+|\s+$/g, '').split("\n")
        var parsed = lines.map(parseJson)
        callback(parsed)
      })
    },

    query: function() {
      return OphionQuery(this)
    },

    q: function() {
      return OphionQuery(this)
    },

    as: function(l) {
      return OphionQuery(this).as(l)
    },

    eq: function(x) {
      return {'eq': value(x)}
    },

    neq: function(x) {
      return {'neq': value(x)}
    },

    gt: function(x) {
      return {'gt': value(x)}
    },

    gte: function(x) {
      return {'gte': value(x)}
    },

    lt: function(x) {
      return {'lt': value(x)}
    },

    lte: function(x) {
      return {'lte': value(x)}
    },

    between: function(lower, upper) {
      return {'between': {'lower': value(lower), 'upper': value(upper)}}
    },

    inside: function(lower, upper) {
      return {'inside': {'lower': value(lower), 'upper': value(upper)}}
    },

    outside: function(lower, upper) {
      return {'outside': {'lower': value(lower), 'upper': value(upper)}}
    },

    within: function(v) {
      return {'within': _.map(v, value)}
    },

    without: function(v) {
      return {'without': _.map(v, value)}
    },

    mark: function(l) {
      return OphionQuery(this).mark(l)
    }
  }
}

module.exports.Ophion = Ophion ;
module.exports.Callsync = callsync.callsync;

File: client/R/R/ophion.R
#' @export
ophion <- function(host) {
  structure(list(), class = "ophion", host = host)
}

#' @export
print.ophion <- function(x) {
  print(paste("host:", attr(x, "host"), sep = " "))
}

#' @export
query <- function(o) {
  class(o) <- c("ophion.query", "ophion")
  attr(o, "query") <- list()
  o
}

#' @export
print.ophion.query <- function(x) {
  print(attr(x, "query"))
}

append.ophion.query <- function(x, values, after = length(x)) {
  q <- attr(x, "query")
  after <- length(q)
  q[[after + 1]] <- values  
  attr(x, "query") <- q
  x
}

#' @export
render <- function(q) {
  jsonlite::toJSON(attr(q, "query"), auto_unbox = T, simplifyVector = F)
}

#' @export
execute <- function(q) {
  body <- render(q)
  print(attributes(q))
  response <- httr::POST(url = paste(attr(q, "host"), "/vertex/query", sep = ""),
                         body = body,
                         httr::content_type_json(),
                         httr::verbose())
  httr::content(response, type="application/json", as="parsed")
}

#' @export
vertex <- function(gid, host) {
  response <- httr::POST(url = paste(host, "/vertex/find/", gid, sep = ""),
                         httr::add_headers("Content-Type"="application/json",
                                           "Accept"="application/json"),
                         httr::verbose())
  httr::content(response)
}

#' @export
incoming <- function(q,  labels=NULL) {
  if (length(labels) == 0) {
    append.ophion.query(q, list("in" = list()))
  } else if (length(args) == 1) {
    append.ophion.query(q, list("in" = list("labels" = list(labels))))
  } else {
    append.ophion.query(q, list("in" = list("labels" = labels)))
  }
}

#' @export
outgoing <- function(q, labels=NULL) {
  if (length(args) == 0) {
    append.ophion.query(q, list("out" = list()))
  } else if (length(args) == 1) {
    append.ophion.query(q, list("out" = list("labels" = list(labels))))
  } else {
    append.ophion.query(q, list("out" = list("labels" = labels)))
  }
}

#' @export
in_edge <- function(q, labels=NULL) {
  if (length(labels) == 0) {
    append.ophion.query(q, list("inEdge" = list()))
  } else if (length(args) == 1) {
    append.ophion.query(q, list("inEdge" = list("labels" = list(labels))))
  } else {
    append.ophion.query(q, list("inEdge" = list("labels" = labels)))
  }
}

#' @export
out_edge <- function(q, labels=NULL) {
  if (length(labels) == 0) {
    append.ophion.query(q, list("outEdge" = list()))
  } else if (length(args) == 1) {
    append.ophion.query(q, list("outEdge" = list("labels" = list(labels))))
  } else {
    append.ophion.query(q, list("outEdge" = list("labels" = labels)))
  }
}

#' @export
in_vertex <- function(q) {
  append.ophion.query(q, list("inVertex" = TRUE))
}

#' @export
out_vertex <- function(q) {
  append.ophion.query(q, list("outVertex" = TRUE))
}

#' @export
mark <- function(q, label) {
  if (is.list(label)) {
    append.ophion.query(q, list("as" = list("labels" = label)))
  } else {
    append.ophion.query(q, list("as" = list("labels" = list(label))))
  }
}

#' @export
select <- function(q, labels) {
  if (length(labels) <= 1) {
    append.ophion.query(q, list("select" = list("labels" = list(labels))))
  } else {
    append.ophion.query(q, list("select" = list("labels" = labels)))
  }
}

#' @export
by <- function(q, label) {
  append.ophion.query(q, list("by" = list("key" = label)))
}

#' @export
label <- function(q) {
  append.ophion.query(q, list("by" = c()))
}

#' @export
values <- function(q, labels) {
  if (length(labels) <= 1) {
    append.ophion.query(q, list("values" = list("labels" = list(labels))))
  } else {
    append.ophion.query(q, list("values" = list("labels" = labels)))
  }
}

#' @export
limit <- function(q, l) {
  append.ophion.query(q, list("limit" = l))
}

#' @export
order <- function(q, key, asc) {
  append.ophion.query(q, list("order" = list("key" = key, "ascending" = asc)))
}

#' @export
range <- function(q, begin, end) {
  append.ophion.query(q, list("range" = list("lower" = begin, "upper" = end)))
}

#' @export
count <- function(q) {
  append.ophion.query(q, list("count" = TRUE))
}

#' @export
dedup <- function(q) {
  append.ophion.query(q, list("dedup" = list()))
}

#' @export
path <- function(q) {
  append.ophion.query(q, list("path": TRUE))
}

#' @export
aggregate <- function(q, label) {
  append.ophion.query(q, list("aggregate" = label))
}

#' @export
group <- function(q, bys) {
  append.ophion.query(q, list("group" = list("bys" = lapply(bys, function(x){ list("key" = x) }))))
}

#' @export
group_count <- function(q, label) {
  if (length(label) == 0) {
    append.ophion.query(q, list("groupCount", c()))
  } else {
    append.ophion.query(q, list("groupCount", list("key" = label)))
  }
}

wrap_value <- function(value) {
  v <- value

  if (length(value) > 1) {
    v <- lapply(value, wrap_value)
  } else if (is.numeric(value)) {
    if (value == round(value)) {
      v <-  list("n" = value)
    } else {
      v <-  list("r" = value)
    }

  } else if (is.character(value)) {
    v = list("s" = value)
  }

  v
}

#' @export
satisfies <- function(q, condition) {
   append.ophion.query(q, list("is" = wrap_value(condition)))
}

#' @export
has <- function(q, key, value= NULL) {
  outer <-  list("key" = key)
  if (!is.null(value)) {
    v <-  wrap_value(value)
    if (is.list(value)) {
      outer <- append(outer, list("condition" = v))
    } else {
      outer <- append(outer, list("value" = v))
    }
  }
  append.ophion.query(q, list("has" = outer))
}

#' @export
has_not <- function(q, key) {
  append.ophion.query(q, list("hasNot" = key))
}

#' @export
match <- function(q, queries) {
  if (length(queries) == 1) {
    queries <- list(queries)
  }
  append.ophion.query(q, list("match", list("queries" = queries)))
}

#' @export
search_vertex <- function(q, search, term=NULL) {
  if (is.null(term)) {
    opts <- list("search" = args[1])
  } else {
    opts <- list("term" = args[1], "search" = args[2])
  }
  append.ophion.query(q, list("searchVertex" = opts))
}

#' @export
search_edge <- function(q,  search, term=NULL) {
  if (is.null(term)) {
    opts <- list("search" = args[1])
  } else {
    opts <- list("term" = args[1], "search" = args[2])
  }
  append.ophion.query(q, list("searchEdge" = opts))
}
